### PR TITLE
added constraints_list for CartesianProduct

### LINKED ISSUE #753: constraints_list for CartesianProduct(Array) of polytopes
Assume that `X` has constraint x <= 1 and `Y` has constraint 3⋅x >= 2 (accessed using `constraints_list`). Then `X × Y` has constraints x1 <= 1 and 3⋅x2 >= 2.

This can be elegantly and efficiently implemented using sparse vectors:
```julia
julia> LinearConstraint(sparsevec([1], [1.]), 1.), LinearConstraint(sparsevec([2], [-3.]), -2.)
(LazySets.HalfSpace{Float64}(  [1]  =  1.0, 1.0), LazySets.HalfSpace{Float64}(  [2]  =  -3.0, -2.0))
```

### PR DESCRIPTION
**Pull request by kostakoida**
Closes #753.

### CODE REVIEW

File: src/CartesianProduct.jl [hunk 1]
```diff
@@ -151,6 +151,39 @@ function ∈(x::AbstractVector{<:Real}, cp::CartesianProduct)::Bool
...
+    clist_low = (constraints_list(cp.X), constraints_list(cp.Y))
+
+    clist = Vector{LinearConstraint{N}}()
+    m = length(clist_low[1]) + length(clist_low[2])
+    sizehint!(clist, m)
+    prev_step = 1
+    # create high-dimensional constraints list
+    for X in clist_low
+        for constr in X
+            new_constr = LinearConstraint(sparsevec(prev_step : (dim(constr) + prev_step-1), constr.a), constr.b)
```
Comment (schillic) [inline src/CartesianProduct.jl]: The indices are a constant vector for each `X`, so I would pull it out of the inner loop:
```julia
if !isempty(X)
    indices = prev_step : (dim(X[1]) + prev_step-1)
end
for constr in X
    new_constr = LinearConstraint(sparsevec(indices, constr.a), constr.b)
...

File: src/CartesianProduct.jl [hunk 2]
```diff
@@ -151,6 +151,39 @@ function ∈(x::AbstractVector{<:Real}, cp::CartesianProduct)::Bool
...
+function constraints_list(cp::CartesianProduct{N})::Vector{LinearConstraint{N}} where N<:Real
+    # collect low-dimensional constraints lists
+    clist_low = (constraints_list(cp.X), constraints_list(cp.Y))
+
+    clist = Vector{LinearConstraint{N}}()
+    m = length(clist_low[1]) + length(clist_low[2])
+    sizehint!(clist, m)
+    prev_step = 1
+    # create high-dimensional constraints list
+    for X in clist_low
```
Comment (schillic) [inline src/CartesianProduct.jl]: Maybe not use `X`, we usually use that for sets.

File: src/CartesianProduct.jl [hunk 3]
```diff
@@ -353,6 +370,53 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+    clist = Vector{LinearConstraint{N}}()
+    sizehint!(clist, n)
+    prev_step = 1
+    # create high-dimensional constraints list
+    for c_low in clist_low
+        if !isempty(c_low)
+            indices = prev_step : (dim(c_low[1]) + prev_step - 1)
+        end
+        for constr in c_low
+            new_constr = LinearConstraint(sparsevec(prev_step : (dim(constr) + prev_step-1), constr.a), constr.b)
```
Comment (schillic) [inline src/CartesianProduct.jl]: Use the `indices` you defined above :smile:

File: src/CartesianProduct.jl [hunk 4]
```diff
@@ -353,6 +370,53 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+### Input
+
+- `cpa` -- Cartesian product
+
+### Output
+
+A list of constraints.
+
+"""
+function constraints_list(cpa::CartesianProductArray{N})::Vector{LinearConstraint{N}} where N<:Real
```
Comment (mforets) [inline src/CartesianProduct.jl]: I think we should better use the interface `constraints_list(cpa::CartesianProductArray{N, <:AbstractPolytope}` given the recent introduction of `constraints_list` list method [for the whole interface](https://github.com/JuliaReach/LazySets.jl/pull/734).

There is a drawback  in this approach (but i don't have a clear picture how to make this play nicely with interfaces), in that the interface restriction i propose will make `constraints_list` of, say, the lazy linear map of a polytope not work. But first we should add `constraints_list(::LazySets.LinearMap{Float64,LazySets.HPolytope{Float64},Float64,Array{Float64,2}})` anyways.

File: src/CartesianProduct.jl [hunk 5]
```diff
@@ -353,6 +370,53 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+    # collect low-dimensional constraints lists
+    clist_low = []
```
Comment (mforets) [inline src/CartesianProduct.jl]: `clist_low = []` is an array of type `Any`. Passing the information known to the compiler will improve efficiency, basically a `clist_low = Vector{LinearConstraint{N}}()`. See [type-stability performance notes](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#Write-%22type-stable%22-functions-1).

File: src/CartesianProduct.jl [hunk 6]
```diff
@@ -353,6 +370,42 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+
+- `cpa` -- Cartesian product
+
+### Output
+
+A list of constraints.
+
+"""
+function constraints_list(cpa::CartesianProductArray{N, <:AbstractPolytope})::Vector{LinearConstraint{N}} where N<:Real
+    # collect low-dimensional constraints lists
```
Comment (schillic) [inline src/CartesianProduct.jl]: redundant now

File: src/CartesianProduct.jl [hunk 7]
```diff
@@ -353,6 +370,42 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+Return the list of constraints of a (polytopic) Cartesian product.
+
+### Input
+
+- `cpa` -- Cartesian product
+
+### Output
+
+A list of constraints.
+
```
Comment (schillic) [inline src/CartesianProduct.jl]: nit-picky: empty line

File: src/CartesianProduct.jl [hunk 8]
```diff
@@ -151,6 +151,23 @@ function ∈(x::AbstractVector{<:Real}, cp::CartesianProduct)::Bool
            ∈(view(x, n1+1:length(x)), cp.Y)
 end
 
+"""
+    constraints_list(cp::CartesianProduct{N})::Vector{LinearConstraint{N}} where N<:Real
```
Comment (schillic) [inline src/CartesianProduct.jl]: update (copy the function header here (without `function`))

File: src/CartesianProduct.jl [hunk 9]
```diff
@@ -353,6 +370,42 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
     return true
 end
 
+"""
+    constraints_list(cpa::CartesianProductArray{N})::Vector{LinearConstraint{N}} where N<:Real
```
Comment (schillic) [inline src/CartesianProduct.jl]: update

File: src/CartesianProduct.jl [hunk 10]
```diff
@@ -353,6 +370,40 @@ function ∈(x::AbstractVector{N}, cpa::CartesianProductArray{N, <:LazySet{N}}
...
+- `cpa` -- Cartesian product
+
+### Output
+
+A list of constraints.
+"""
+function constraints_list(cpa::CartesianProductArray{N, <:AbstractPolytope})::Vector{LinearConstraint{N}} where N<:Real
+    c_array = array(cpa)
+    clist = Vector{LinearConstraint{N}}()
+    sizehint!(clist, sum(dim(s_low) for s_low in c_array))
```
Comment (mforets) [inline src/CartesianProduct.jl]: note that `dim(cpa)` exists too

File: src/CartesianProduct.jl [hunk 11]
```diff
@@ -391,10 +391,10 @@ function constraints_list(cpa::CartesianProductArray{N, <:LazySet{N}})::Vector{L
     for c_low in array(cpa)
         c_low_list = constraints_list(c_low)
         if !isempty(c_low_list)
-            indices = prev_step : (dim(c_low_list[1]) + prev_step - 1)
+            indices = collect(prev_step : (dim(c_low_list[1]) + prev_step - 1))
```
Comment (schillic) [inline src/CartesianProduct.jl]: What was bad about the colon notation? It was more efficient...

**Comment by mforets** [comment]
@kostakoida  feel free to add your name in https://juliareach.github.io/LazySets.jl/latest/about.html#Credits-1

**Pull request merged**